### PR TITLE
Feature/Xline Client

### PR DESCRIPTION
**Pull request by bsbds**
Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

    Add a new Xline client crate, offering an external client for Xline.

* what changes does this pull request make?

   `xline-client` crate. The clients are currently placeholders, they will get implemented in other PRs.

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

    No

### CODE REVIEW

File: xline-client/src/clients/auth.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
+// TODO: Remove these when the placeholder is implemented.
+#![allow(dead_code)]
+
+use std::sync::Arc;
+
+use curp::client::Client as CurpClient;
+use tonic::transport::Channel;
+use xline::server::Command;
+
+use crate::AuthService;
```
Comment (rogercloud) [inline xline-client/src/clients/auth.rs:10]: It's a client crate, why do we have a service here?

Comment (bsbds) [reply xline-client/src/clients/auth.rs:10]: The `AuthService` naming came from `tower::Service` trait, which is a abstraction from a request to a respond https://docs.rs/tower/latest/tower/trait.Service.html. Here the `AuthService` is used for token authentication.

File: xline-client/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+            .field("lease", &self.lease)
+            .field("lock", &self.lock)
+            .field("auth", &self.auth)
+            .field("maintenance", &self.maintenance)
+            .field("watch", &self.watch)
+            .field("cluster", &self.cluster)
+            .field("election", &self.election)
+            .finish()
+    }
+}
```
Comment (themanforfree) [inline xline-client/src/lib.rs:202]: Please use `#[derive(Debug)]`

File: xline-client/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+
+impl Client {
+    /// New `Client`
+    ///
+    /// # Errors
+    ///
+    /// If `StandaloneClient::connect` fails.
```
Comment (themanforfree) [inline xline-client/src/lib.rs:209]: Please update comments

File: xline-client/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+    #[inline]
+    pub async fn connect(
+        all_members: HashMap<String, String>,
+        options: ClientOptions,
+    ) -> Result<Self> {
+        let name = String::from("client");
+        let channel = Self::build_channel(&all_members).await?;
+        let curp_client = Arc::new(CurpClient::new(None, all_members, options.curp_timeout).await);
+        let token = None;
```
Comment (themanforfree) [inline xline-client/src/lib.rs:218]: If this PR does not really use token, please add a todo

File: xline-client/src/clients/cluster.rs [hunk 5]
```diff
@@ -0,0 +1,17 @@
+// TODO: Remove these when the placeholder is implemented.
+#![allow(missing_copy_implementations)]
+#![allow(clippy::new_without_default)]
```
Comment (themanforfree) [inline xline-client/src/clients/cluster.rs:3]: this PR does not need these lines, please add them when you need them, same as other files

File: xline-client/src/lib.rs (hunk 2 above)
Comment (bsbds) [reply xline-client/src/lib.rs]: done

File: xline-client/src/lib.rs (hunk 3 above)
Comment (bsbds) [reply xline-client/src/lib.rs]: done

File: xline-client/src/lib.rs (hunk 4 above)
Comment (bsbds) [reply xline-client/src/lib.rs:201]: done

File: xline-client/src/clients/cluster.rs (hunk 5 above)
Comment (bsbds) [reply xline-client/src/clients/cluster.rs:3]: The files act as placeholders. I need these to satisfy the crate clippy lints.

File: xline-client/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+        self.inner.call(request)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[tokio::test]
+    async fn connect() {
```
Comment (Phoenix500526) [inline xline-client/src/lib.rs:394]: Why do we need such a trivial test here? It looks useless.

File: xline-client/src/clients/election.rs [hunk 7]
```diff
@@ -0,0 +1,17 @@
+// TODO: Remove these when the placeholder is implemented.
+#![allow(missing_copy_implementations)]
+#![allow(clippy::new_without_default)]
+
+/// The election client
+#[derive(Clone, Debug)]
+#[non_exhaustive]
+pub struct ElectionClient;
```
Comment (Phoenix500526) [inline xline-client/src/clients/election.rs:8]: Why should we need such an election client? What is it responsible for?  In what situation we will need it?

File: xline-client/src/clients/lease.rs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+/// Client for Lease operations.
+#[derive(Clone, Debug)]
+pub struct LeaseClient {
+    /// Name of the LeaseClient
+    name: String,
+    /// The CURP client
+    curp_client: Arc<CurpClient<Command>>,
+    /// The standalone client
+    standalone_client: xlineapi::LeaseClient<AuthService<Channel>>,
```
Comment (Phoenix500526) [inline xline-client/src/clients/lease.rs:20]: Why should `LeaseClient` be placed in the `xlineapi` crate while the others, like `CurpClient` or `WatchClient`, are in the `xline-client`? By the way, what does the standalone mean? Can we rename it as lease_client?

File: xline-client/src/clients/maintenance.rs [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+use tonic::transport::Channel;
+
+use crate::AuthService;
+
+/// The maintenance client
+#[derive(Clone, Debug)]
+pub struct MaintenanceClient {
+    /// the inner RPC client
+    inner: xlineapi::MaintenanceClient<AuthService<Channel>>,
```
Comment (Phoenix500526) [inline xline-client/src/clients/maintenance.rs:14]: Same as above

File: xline-client/src/clients/watch.rs [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+use tonic::transport::Channel;
+
+use crate::AuthService;
+
+/// The maintenance client
+#[derive(Clone, Debug)]
+pub struct WatchClient {
+    /// The inner RPC client
+    inner: xlineapi::WatchClient<AuthService<Channel>>,
```
Comment (Phoenix500526) [inline xline-client/src/clients/watch.rs:14]: Same as above.

File: xline-client/Cargo.toml [hunk 11]
```diff
@@ -0,0 +1,31 @@
+[package]
+name = "xline-client"
+version = "0.1.0"
+edition = "2021"
+authors = ["DatenLord <dev@datenlord.io>"]
+description = "Client for Xline"
+license = "Apache-2.0"
+readme = "README.md"
+repository = "https://github.com/datenlord/Xline/tree/master/xline-client"
+categories = ["Client"]
```
Comment (Phoenix500526) [inline xline-client/Cargo.toml:10]: How about Command Line Tool here?

File: xline-client/src/lib.rs (hunk 6 above)
Comment (bsbds) [reply xline-client/src/lib.rs:394]: Just test basic functionality creating a client to satisfy the ci coverage. I could remove this if not needed.

File: xline-client/src/clients/election.rs (hunk 7 above)
Comment (bsbds) [reply xline-client/src/clients/election.rs:8]: It came from etcd client https://github.com/etcd-io/etcd/blob/main/client/v3/concurrency/election.go. 
If we are gonna implement this feature in the future.

File: xline-client/src/clients/lease.rs (hunk 8 above)
Comment (bsbds) [reply xline-client/src/clients/lease.rs:20]: Not sure what you mean, the `xlineapi::LeaseClient` and `xlineapi::WatchClient` are generated from the protobuf files. And `CurpClient` is defined in the `curp` crate.

The `standalone` is relative to the `curp` client, which means curp client need to communicate with multiple server while the standalone client only communicate with one at a time.

File: xline-client/Cargo.toml (hunk 11 above)
Comment (bsbds) [reply xline-client/Cargo.toml:10]: This crate is just the client API. The binary command line tool will be implemented in another crate.

File: xline-client/src/clients/lease.rs (hunk 8 above)
Comment (Phoenix500526) [reply xline-client/src/clients/lease.rs:20]: I recommend renaming it as `lease_client` and updating the standalone meaning in its comment.

File: xline-client/src/lib.rs (hunk 6 above)
Comment (bsbds) [reply xline-client/src/lib.rs]: done

File: xline-client/src/clients/lease.rs (hunk 8 above)
Comment (bsbds) [reply xline-client/src/clients/lease.rs]: done

**Review by Phoenix500526 — APPROVED** [review]

**Review by themanforfree — APPROVED** [review]

**Pull request merged**